pkg/auth/multiauthenticator_test.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMultiAuthenticatorWithClientID_GetAuthenticator_Ok(t *testing.T) {
+	// Given.
+	jwtAuth, err := NewJwtAuthenticator(&JwtAuthConfig{
+		SharedSecret:  []byte("my-secret"),
+		UsernameClaim: UsernameClaimTypeSubject,
+	})
+	assert.NoError(t, err)
+	m := NewMultiAuthenticatorWithClientID(map[IssuerWithClientID]Authenticator{IssuerWithClientID{"issuer-1", "1"}: jwtAuth})
+
+	// When.
+	found := m.GetAuthenticator("issuer-1", "1")
+
+	// Then.
+	assert.NotNil(t, found)
+}
+
+func TestMultiAuthenticatorWithClientID_GetAuthenticator_FailNoClientId(t *testing.T) {
+	// Given.
+	jwtAuth, err := NewJwtAuthenticator(&JwtAuthConfig{
+		SharedSecret:  []byte("my-secret"),
+		UsernameClaim: UsernameClaimTypeSubject,
+	})
+	assert.NoError(t, err)
+	m := NewMultiAuthenticatorWithClientID(map[IssuerWithClientID]Authenticator{
+		IssuerWithClientID{"issuer-1", "1"}: jwtAuth,
+	})
+
+	// When.
+	found := m.GetAuthenticator("issuer-1", "2")
+
+	// Then.
+	assert.Nil(t, found)
+}
+
+func TestMultiAuthenticatorWithClientID_GetAuthenticator_FailNoIssuer(t *testing.T) {
+	// Given.
+	m := NewMultiAuthenticatorWithClientID(map[IssuerWithClientID]Authenticator{})
+
+	// When.
+	found := m.GetAuthenticator("issuer-1", "1")
+
+	// Then.
+	assert.Nil(t, found)
+}
+
+func TestMultiAuthenticatorWithClientID_AuthenticateToken_Ok(t *testing.T) {
+	// Given.
+	jwtAuth1, err := NewJwtAuthenticator(&JwtAuthConfig{
+		SharedSecret:  []byte("my-secret-1"),
+		UsernameClaim: UsernameClaimTypeSubject,
+	})
+	assert.NoError(t, err)
+
+	m := NewMultiAuthenticatorWithClientID(map[IssuerWithClientID]Authenticator{
+		IssuerWithClientID{"issuer-1", "1"}: jwtAuth1,
+		IssuerWithClientID{"issuer-1", "2"}: jwtAuth1,
+	})
+
+	rawToken, err := Token(&Claims{
+		Audience: "1",
+		Issuer:   "issuer-1",
+		Email:    "my@email.com",
+		Name:     "my-name",
+		Subject:  "my-sub",
+		Roles:    []string{"tester"},
+	}, &Signature{
+		Type: jwt.SigningMethodHS256,
+		Key:  []byte("my-secret-1"),
+	}, &Options{
+		Expiration: time.Now().Add(time.Minute * 10).Unix(),
+	})
+	assert.NoError(t, err)
+	assert.NotEmpty(t, rawToken)
+
+	// When.
+	authenticateClaims, err := m.AuthenticateToken(context.Background(), rawToken)
+
+	// Then.
+	assert.NoError(t, err)
+	assert.Equal(t, "issuer-1", authenticateClaims.Issuer)
+	assert.Equal(t, "my@email.com", authenticateClaims.Email)
+	assert.Equal(t, "my-name", authenticateClaims.Name)
+	assert.Equal(t, []string{"tester"}, authenticateClaims.Roles)
+}
+
+func TestMultiAuthenticatorWithClientID_AuthenticateTokenMultiAud_Ok(t *testing.T) {
+	// Given.
+	jwtAuth1, err := NewJwtAuthenticator(&JwtAuthConfig{
+		SharedSecret:  []byte("my-secret-1"),
+		UsernameClaim: UsernameClaimTypeSubject,
+	})
+	assert.NoError(t, err)
+
+	m := NewMultiAuthenticatorWithClientID(map[IssuerWithClientID]Authenticator{
+		IssuerWithClientID{"issuer-1", ""}:          jwtAuth1,
+		IssuerWithClientID{"issuer-1", "clientID2"}: jwtAuth1,
```

Comment:
I don't see the point of this setup in practice. Feels like we should not allow it. The first authenticator with empty cilentID will always produce a match for that issuer or does Contains("") work differently than what I am thinking? Which authenticator will get used is non-deterministic since order of iteration through the map is can be random.